### PR TITLE
aggregate cluster roles

### PR DESCRIPTION
**Pull request by deads2k**
xref https://github.com/kubernetes/community/pull/1219 https://github.com/kubernetes/features/issues/502

This is a pull with API types, a controller, and a demonstration of how to move admin, edit, and view.  Once we agree on the shape, I'll 

I added 
```yaml
aggregationRule:
  clusterRoleSelectors:
  - matchLabels:
      rbac.authorization.k8s.io/aggregate-to-admin: true
```
to the `ClusterRole`.  A controller then goes and gathers all the matching ClusterRoles and sets the `rules` to the union of matching cluster roles.

@kubernetes/sig-auth-pr-reviews 

```release-note
RBAC ClusterRoles can now select other roles to aggregate
```

### CODE REVIEW

**Comment by deads2k** [comment]
Opened https://github.com/kubernetes/community/pull/1219 in case people would rather read this in doc form

File: pkg/registry/rbac/reconciliation/reconcile_role.go [hunk 1]
```diff
@@ -195,6 +202,20 @@ func computeReconciledRole(existing, expected RuleOwner, removeExtraPermissions
 		result.Operation = ReconcileUpdate
 	}
 
+	switch {
+	case existing.GetAggregationRule() == nil && expected.GetAggregationRule() == nil:
+	case existing.GetAggregationRule() != nil && expected.GetAggregationRule() == nil:
+		result.ReplacementAggregationRule = nil
+		result.Role.SetSetAggregationRule(nil)
```
Comment (liggitt) [inline pkg/registry/rbac/reconciliation/reconcile_role.go]: `SetSet`?

File: pkg/registry/rbac/reconciliation/reconcile_role.go [hunk 2]
```diff
@@ -195,6 +202,20 @@ func computeReconciledRole(existing, expected RuleOwner, removeExtraPermissions
...
+		result.Operation = ReconcileUpdate
+	case existing.GetAggregationRule() == nil && expected.GetAggregationRule() != nil,
+		// it is uncommon to have aggregation rules, just always reconcile these
+		existing.GetAggregationRule() != nil && expected.GetAggregationRule() != nil:
```
Comment (liggitt) [inline pkg/registry/rbac/reconciliation/reconcile_role.go]: when both are non-nil, I'm wary of stomping what was already there

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go [hunk 3]
```diff
@@ -176,6 +176,30 @@ func ClusterRoles() []rbac.ClusterRole {
...
+			// a role for namespace level viewing.  It grants Read-only access to non-escalating resources in
+			// a namespace.
+			ObjectMeta: metav1.ObjectMeta{Name: "view"},
+			AggregationRule: &rbac.AggregationRule{
+				Selectors: []metav1.LabelSelector{{MatchLabels: map[string]string{"rbac.authorization.k8s.io/aggregate-to-clusterrole-name": "view"}}},
+			},
+		},
+		{
+			// a role for a namespace level admin.  It is `edit` plus the power to grant permissions to other users.
+			ObjectMeta: metav1.ObjectMeta{Name: "admin-to-aggregate", Labels: map[string]string{"rbac.authorization.k8s.io/aggregate-to-clusterrole-name": "admin"}},
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: `system:aggregate-to-admin`, etc?

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go [hunk 4]
```diff
@@ -176,6 +176,30 @@ func ClusterRoles() []rbac.ClusterRole {
 		{
 			// a role for a namespace level admin.  It is `edit` plus the power to grant permissions to other users.
 			ObjectMeta: metav1.ObjectMeta{Name: "admin"},
+			AggregationRule: &rbac.AggregationRule{
+				Selectors: []metav1.LabelSelector{{MatchLabels: map[string]string{"rbac.authorization.k8s.io/aggregate-to-clusterrole-name": "admin"}}},
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: `.../aggregate-to-admin": "true"`, etc

File: staging/src/k8s.io/api/rbac/v1/types.go [hunk 5]
```diff
@@ -170,6 +170,18 @@ type ClusterRole struct {
 
 	// Rules holds all the PolicyRules for this ClusterRole
 	Rules []PolicyRule `json:"rules" protobuf:"bytes,2,rep,name=rules"`
+
+	// AggregationRule is an optional field that describes how to build the Rules for this ClusterRole.
+	// If AggregationRule is set, then the Rules are controller managed and direct changes to Rules will be
+	// stomped by the controller.
+	AggregationRule *AggregationRule `json:"aggregationRule" protobuf:"bytes,3,opt,name=aggregationRule"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/rbac/v1/types.go]: omitempty

File: pkg/registry/rbac/reconciliation/reconcile_role.go (hunk 2 above)
Comment (deads2k) [reply pkg/registry/rbac/reconciliation/reconcile_role.go]: > when both are non-nil, I'm wary of stomping what was already there

I did make it a slice.  Would you like to combine them?

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go (hunk 3 above)
Comment (deads2k) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: > system:aggregate-to-admin, etc? we want as few unnamespaced roles as possible

`system:built-in-admin`?  Split into one per group and have `system:core-admin`, `system:extensions-admin`?

File: pkg/registry/rbac/reconciliation/reconcile_role.go (hunk 2 above)
Comment (liggitt) [reply pkg/registry/rbac/reconciliation/reconcile_role.go]: I think I'd expect the default ones to be added if not covered by one already in the slice. auto-reconciliation is strictly additive today

**Comment by deads2k** [comment]
comments addressed, controller test added, escalation check in place.

**Comment by deads2k** [comment]
now with passing linting.

File: hack/.golint_failures [hunk 6]
```diff
@@ -144,6 +144,7 @@ pkg/controller/certificates
 pkg/controller/certificates/approver
 pkg/controller/certificates/signer
 pkg/controller/cloud
+pkg/controller/clusterroleaggregation
```
Comment (deads2k) [inline hack/.golint_failures]: I added this because I had one failure about `pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go:42:6: type name will be used as clusterroleaggregation.ClusterRoleAggregationController by other packages, and that stutters; consider calling this Controller`, but I think the specific name is substantially better given that **everything** is a controller in the files this will be called from.

**Comment by deads2k** [comment]
@enj ptal

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 7]
```diff
@@ -310,3 +317,37 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
...
 }
+
+// clusterRolesToAggregate maps from previous clusterrole name to the new clusterrole name
+var clusterRolesToAggregate = map[string]string{
+	"admin": "system:aggregate-to-admin",
+	"edit":  "system:aggregate-to-edit",
+	"view":  "system:aggregate-to-view",
+}
+
+// primeAggregatedClusterRoles
```
Comment (ericchiang) [inline pkg/registry/rbac/rest/storage_rbac.go]: nit: godoc not complete?

**Review by ericchiang — COMMENTED** [review]
The reconciliation, bootstrap policy changes, and storage migration are clever. Upgrade/downgrade path ended up really clean.

**Review by enj — CHANGES_REQUESTED** [review]

**Comment by enj** [comment]
@deads2k @liggitt https://github.com/kubernetes/kubernetes/pull/54005#discussion_r148404641 is my main concern here.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go [hunk 8]
```diff
@@ -176,6 +176,30 @@ func ClusterRoles() []rbac.ClusterRole {
 		{
 			// a role for a namespace level admin.  It is `edit` plus the power to grant permissions to other users.
 			ObjectMeta: metav1.ObjectMeta{Name: "admin"},
+			AggregationRule: &rbac.AggregationRule{
```
[the remark this answers is not in the document]
Comment (deads2k) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: > I know we have tests that make sure that admin covers edit and so on. The counter argument would be "I do not want to piece together 3 thing to understand what admin means" - granted the YAML files should cover that case (but reading the union of a bunch of rules is probably harder).

Jordan and I debated how far to decompose this.  The two of us did not reach consensus, so we're pausing here for now.  Even if you want to do that, you can do it two way: admin can gather admins, edits, view or the constituent roles and tag themselves multiple ways.

File: pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", clusterRole, err))
+			return
+		}
+		c.queue.Add(key)
+	}
+}
+
+func (c *ClusterRoleAggregationController) delete(uncast interface{}) {
+	obj, ok := uncast.(*rbacv1.ClusterRole)
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: > Do controllers normally work on external types?

Yeah.  And then many of them have to convert to other types through the internal hub.  So many battles, so little time.

File: pkg/registry/rbac/clusterrole/policybased/storage.go [hunk 10]
```diff
@@ -68,8 +80,16 @@ func (s *Storage) Update(ctx genericapirequest.Context, name string, obj rest.Up
 
 		rules := clusterRole.Rules
 		if err := rbacregistryvalidation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
-			return nil, errors.NewForbidden(groupResource, clusterRole.Name, err)
+			return nil, apierrors.NewForbidden(groupResource, clusterRole.Name, err)
+		}
+		// to change the aggregation rule, since it can gather anything and prevent tightening, requires * on *.*
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/registry/rbac/clusterrole/policybased/storage.go]: > Don't you also need to check for the creation or update of a cluster role that will then get aggregated into a different cluster role via an AggregationRule?

When you modify a given ClusterRole, you have to all the permissions you're trying to add.  That means that you have the permission that will eventually be added to the aggregated cluster role and thus added to other users.  We're trying to protect against escalation and I think this does that.  

In the other direction, we have to protect against, "aggregate cluster admin into me".  Forcing a super-user does that.



> The situation I am thinking of is:
> 
> Cluster Role A { aggregates Cluster Role B and Cluster Role C via label D }
> Cluster Role B (has label D) = Rule 1
> Cluster Role C (has label D) = Rule 2
> 
> I only have Rule 2, so I can only pass the escalation check for C. I remove the label from C. This causes A to be tightened (which I normally could not do because I do not have Rule 1). The reverse is also true (if I can create or update a cluster role with label D, I can cause A to be updated even if I would normally not be able to do that).

We're primarily focused on escalation cases. The idea being, "if you have the power to mess with some roles, we need to prevent you from adding more permissions to yourself".  Are you worried about a case where a cluster-admin trusts someone to control what a particular clusterrole can do, but doesn't want that person to be able to control bindings of that role (that's what it sounds like you're describing).  If the role-admin already has the power to perform an action, he could just as easily say "write to this proxy and I'll perform actions for you".  What are you really preventing?

Modifying cluster-roles is already pretty privileged and I'm not seeing how it allows a user to escalate himself (which is what we're trying to protect against).  Given how people would handle this today (secondary binding or direct modification), the power seems the same to me.

File: pkg/registry/rbac/clusterrole/policybased/storage.go [hunk 11]
```diff
@@ -68,10 +80,21 @@ func (s *Storage) Update(ctx genericapirequest.Context, name string, obj rest.Up
...
 		if err := rbacregistryvalidation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
-			return nil, errors.NewForbidden(groupResource, clusterRole.Name, err)
+			return nil, apierrors.NewForbidden(groupResource, clusterRole.Name, err)
 		}
+		// to change the aggregation rule, since it can gather anything and prevent tightening, requires * on *.*
+		if hasAggregationRule(clusterRole) || hasAggregationRule(oldClusterRole) {
+			if err := rbacregistryvalidation.ConfirmNoEscalation(ctx, s.ruleResolver, fullAuthority); err != nil {
+				return nil, apierrors.NewForbidden(groupResource, clusterRole.Name, errors.New("must have cluster-admin privileges to use the aggregationRule"))
+			}
+		}
```
Comment (liggitt) [inline pkg/registry/rbac/clusterrole/policybased/storage.go]: if we have aggregation rules and validate fullAuthority, we don't need the detailed covers check on the role rules

File: pkg/registry/rbac/clusterrole/policybased/storage.go [hunk 12]
```diff
@@ -40,6 +42,8 @@ func NewStorage(s rest.StandardStorage, ruleResolver rbacregistryvalidation.Auth
 	return &Storage{s, ruleResolver}
 }
 
+var fullAuthority = []rbac.PolicyRule{rbac.NewRule("*").Groups("*").Resources("*").RuleOrDie()}
```
Comment (liggitt) [inline pkg/registry/rbac/clusterrole/policybased/storage.go]: also include `*` on `nonResourceURLs:["*"]`

File: staging/src/k8s.io/api/rbac/v1beta1/types.go [hunk 13]
```diff
@@ -170,7 +170,21 @@ type ClusterRole struct {
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
 	// Rules holds all the PolicyRules for this ClusterRole
+
```
Comment (liggitt) [inline staging/src/k8s.io/api/rbac/v1beta1/types.go]: remove spurious whitespace (also dropped generated description in openapi)

File: pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+	// graph, just queue up every thing each time.  This allows errors to be selectively retried if there
+	// is a problem updating a single role
+	allClusterRoles, err := c.clusterRoleLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't list all objects %v", err))
+		return
+	}
+	for _, clusterRole := range allClusterRoles {
+		// only queue ones that we may need to aggregate
+		if clusterRole.AggregationRule == nil {
```
Comment (liggitt) [inline pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: is it an error to have non-nil aggregationRule and zero clusterRoleSelectors?

Comment (deads2k) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: > is it an error to have non-nil aggregationRule and zero clusterRoleSelectors?

The controller won't do anything with it.  I don't think I straight out made it an error though.  I don't feel strongly.

File: pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+		clusterRoles, err := c.clusterRoleLister.List(runtimeLabelSelector)
+		if err != nil {
+			return err
+		}
+		sort.Sort(byName(clusterRoles))
+
+		for i := range clusterRoles {
+			if clusterRoles[i].Name == sharedClusterRole.Name {
+				continue
+			}
```
Comment (liggitt) [inline pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: also track if we already aggregated this role by name and skip the rule-by-rule checking?

Comment (deads2k) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: > also track if we already aggregated this role by name and skip the rule-by-rule checking?

I'd rather avoid the extra tracking in this loop.  The rules would already be reject below.  Remember, we have o(10-ish) aggregated roles and they're usually steady state.  It's not like its a high throughput spot.

File: pkg/registry/rbac/reconciliation/reconcile_role_test.go [hunk 16]
```diff
@@ -268,8 +272,102 @@ func TestComputeReconciledRole(t *testing.T) {
 			t.Errorf("%s: Expected\n\t%v\ngot\n\t%v", k, tc.expectedReconciliationNeeded, reconciliationNeeded)
 			continue
 		}
+		result.Role.(ClusterRoleRuleOwner).ClusterRole.AggregationRule = nil
```
Comment (liggitt) [inline pkg/registry/rbac/reconciliation/reconcile_role_test.go]: why was this change needed? I would have expected it to be nil already

File: pkg/registry/rbac/reconciliation/reconcile_role_test.go [hunk 17]
```diff
@@ -268,8 +272,102 @@ func TestComputeReconciledRole(t *testing.T) {
...
+		expectedReconciledRole       *rbac.ClusterRole
+		expectedReconciliationNeeded bool
+	}{
+		"empty": {
+			expectedRole:           aggregatedRole(&rbac.AggregationRule{}),
+			actualRole:             aggregatedRole(nil),
+			removeExtraPermissions: true,
+
+			expectedReconciledRole:       nil,
+			expectedReconciliationNeeded: false,
```
Comment (liggitt) [inline pkg/registry/rbac/reconciliation/reconcile_role_test.go]: if expected has a non-nil aggregation rule, and existing has none, that's a meaningful difference today (one would trigger aggregation and the other would not) and should trigger reconciliation

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 18]
```diff
@@ -310,3 +317,38 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
 func (p RESTStorageProvider) GroupName() string {
 	return rbac.GroupName
 }
+
+// ClusterRolesToAggregate maps from previous clusterrole name to the new clusterrole name
+var ClusterRolesToAggregate = map[string]string{
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go]: agree the map of role name to aggregated role name should come from bootstrap policy, not be inlined here

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 19]
```diff
@@ -310,3 +317,38 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
...
+		}
+
+		existingRole, err := clusterRoleClient.ClusterRoles().Get(oldName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		existingRole.Name = newName
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go]: probably worth logging the create attempt here

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 20]
```diff
@@ -310,3 +317,38 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
...
+		if _, err := clusterRoleClient.ClusterRoles().Create(existingRole); err != nil {
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go]: should tolerate an AlreadyExists error in case HA masters race startup

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go [hunk 21]
```diff
@@ -69,6 +69,13 @@ func buildControllerRoles() ([]rbac.ClusterRole, []rbac.ClusterRoleBinding) {
 		},
 	})
 	addControllerRole(&controllerRoles, &controllerRoleBindings, rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "clusterrole-aggregation-controller"},
+		Rules: []rbac.PolicyRule{
+			// this controller must have full permissions to allow it to mutate any role in any way
+			rbac.NewRule("*").Groups("*").Resources("*").RuleOrDie(),
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go]: and non-resource urls

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 22]
```diff
@@ -176,6 +177,12 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
 				return false, nil
 			}
 
+			// if the new cluster roles to aggregate do not yet exist, then we need to copy the old roles if they don't exist
+			// in new locations
+			if err := primeAggregatedClusterRoles(ClusterRolesToAggregate, clientset); err != nil {
+				return false, err
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go]: call HandleError and return `false, nil`... we want to trigger retry

File: pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go (hunk 14 above)
Comment (liggitt) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: > The controller won't do anything with it.

I don't think that's true... looks like syncClusterRole would handle it, select zero roles, and empty out `rules`

Comment (deads2k) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: > I don't think that's true... looks like syncClusterRole would handle it, select zero roles, and empty out rules

Ok.  That doesn't bother me either.  You have a preference against it or just observing?

Comment (liggitt) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: I think I'm ok with it as well, but reconciliation needs to be fixed

File: pkg/registry/rbac/clusterrole/policybased/storage.go (hunk 11 above)
Comment (deads2k) [reply pkg/registry/rbac/clusterrole/policybased/storage.go]: > if we have aggregation rules and validate fullAuthority, we don't need the detailed covers check on the role rules

I'd really rather put it in an else branch branch unless you feel strongly.  It's not like these are hit often and the unconditional check is very easy to read and reason about.  Makes me feel safe.

File: pkg/registry/rbac/reconciliation/reconcile_role_test.go (hunk 16 above)
Comment (deads2k) [reply pkg/registry/rbac/reconciliation/reconcile_role_test.go]: > why was this change needed? I would have expected it to be nil already

removing.

File: pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go (hunk 14 above)
Comment (deads2k) [reply pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go]: > is it an error to have non-nil aggregationRule and zero clusterRoleSelectors?

Turned out to be less work to just tighten validation.  I started there.  We could loosen it later.

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 23]
```diff
@@ -179,8 +182,8 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
 
 			// if the new cluster roles to aggregate do not yet exist, then we need to copy the old roles if they don't exist
 			// in new locations
-			if err := primeAggregatedClusterRoles(ClusterRolesToAggregate, clientset); err != nil {
-				return false, err
+			if err := primeAggregatedClusterRoles(p.ClusterRolesToAggregate, clientset); err != nil {
+				return false, nil
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go]: log error with `HandleError`

**Comment by liggitt** [comment]
one logging nit, then squash
/approve
